pkg/recommendedversion/recommended_version.go

```diff
@@ -0,0 +1,263 @@
+func SortRecommendedVersionsDescending(recommendedVersionStr string) ([]string, error) {
+	// The value is in the form "v1.2.1,v1.1.0,v0.90.1"
+	// which is a comma separated list of recommended versions for each minor version of the CLI.
+	recommendedArray := strings.Split(recommendedVersionStr, ",")
+
+	// Trim any spaces around the version strings and remove duplicates
+	recommendedVersions := make([]string, 0, len(recommendedArray))
+	alreadyPresent := make(map[string]bool)
+	for _, newVersion := range recommendedArray {
+		trimmedVersion := strings.TrimSpace(newVersion)
+		if trimmedVersion != "" && !alreadyPresent[trimmedVersion] {
+			recommendedVersions = append(recommendedVersions, trimmedVersion)
+			alreadyPresent[trimmedVersion] = true
+		}
+	}
+
+	// Now sort the versions, then reverse the order
+	err := utils.SortVersions(recommendedVersions)
+	if err != nil {
+		return nil, err
+	}
+
+	// Reverse the order so it is descending
+	for i := len(recommendedVersions)/2 - 1; i >= 0; i-- {
+		opp := len(recommendedVersions) - 1 - i
+		recommendedVersions[i], recommendedVersions[opp] = recommendedVersions[opp], recommendedVersions[i]
+	}
+	return recommendedVersions, err
+}
+
+func getRecommendationDelayValue() int {
+	delay := recommendedVersionCheckDelaySeconds
+	delayOverride := os.Getenv(constants.ConfigVariableRecommendVersionDelay)
+	if delayOverride != "" {
+		delayOverrideValue, err := strconv.Atoi(delayOverride)
+		if err == nil && delayOverrideValue >= 0 {
+			delay = delayOverrideValue
+		}
+	}
+	return delay
+}
+
+func shouldCheckVersion() bool {
+	delay := getRecommendationDelayValue()
+	if delay == 0 {
+		// The user has disabled the version check
+		return false
+	}
+
+	// Get the last time the version check was done
+	lastCheck := datastore.GetDataStoreValue(dataStoreLastVersionCheckKey)
+	if lastCheck == nil {
+		return true
+	}
+
+	lastCheckTime, ok := lastCheck.(time.Time)
+	if !ok {
+		return true
+	}
+
+	return time.Since(lastCheckTime) > time.Duration(delay)*time.Second
+}
+
+func printVersionRecommendations(writer io.Writer, currentVersion, major, minor, patch string) {
+	if major == "" && minor == "" && patch == "" {
+		// The current version is the best recommended version
+		return
+	}
+
+	// Put a delimiter before this notification so the user
+	// can see it is not part of the command output
+	fmt.Fprintln(writer, "\n==")
+
+	if utils.IsNewVersion(currentVersion, major) || utils.IsNewVersion(currentVersion, minor) || utils.IsNewVersion(currentVersion, patch) {
+		fmt.Fprintf(writer, "WARNING: Due to a problem it is recommended not to use the current version: %s.\n", currentVersion)
+		fmt.Fprintln(writer, "Please use a recommended version:")
+	} else {
+		fmt.Fprintf(writer, "Note: A new version of the Tanzu CLI is available. You are at version: %s.\n", currentVersion)
+		fmt.Fprintln(writer, "To benefit from the latest security and features, please update to a recommended version:")
+	}
+
+	if major != "" {
```

Comment:
**Thought:** Should we be showing this notification when we come out with a newer major version? The major version is a big change and with the breaking changes, it is unlikely that the user will be upgrading the CLI to the new Major version. In that case, the user will need to specify `TANZU_CLI_RECOMMEND_VERSION_DELAY_SECONDS` to a very high value to avoid this additional message.

Considering that, I am thinking whether we should just be showing this warning for `minor` and `patch` version upgrades?

Reply:
Another good point.

I've seen many tool print things like "mytool v4.0 is coming soon! look for the release", so I thought showing the MAJOR version would be nice.  But you are right that if I don't want the MAJOR Version I will be annoyed that the CLI keeps mentioning it.

I think MAJORs are so rare that we can remove the printout.